### ISSUE TITLE
track training request applicant scores

### ISSUE THREAD

**Issue by ErinBecker** [swcarpentry/amy#1327]
When an individual not affiliated with a member institution wants to become an Instructor, they fill out the [application form](https://amy.software-carpentry.org/forms/request_training/). Every few weeks(?) @karenword downloads the new applications and she and other Trainers score them according to the [rubric](https://github.com/carpentries/instructor-training/blob/gh-pages/files/rubric.md). She stores scores in an Excel sheet and manually emails the top N-scored applicants when seats open in an instructor training event. This workflow is sub-optimal for a couple of reasons:

1) @karenword has to use Excel functions to avoid overwriting applicant scores in the Excel sheet when she pulls new applicants. 
2) If an applicant contacts team@ asking about their application, no one but @karenword has access to the spreadsheet to see what their status is.
3) Most of the rubric is auto-scorable, but @karenword and the other Trainers do it by hand. 
4) We can't automate email generation to applicants based on the number of seats available in trainings. 

It would be very useful to have a field in the "trainee" record in AMY that we could use to record their application score. I propose three fields:
1) `score-auto`: generated automatically using the rubric
2) `score-manual`: added manually by @karenword
3) `score-total`: the sum of (1) and (2)

When an instructor training event is created, it has a start date field. When the start date is 29 days away, can AMY calculate the number of seats remaining (25 - number of learners associated with the event) and generate an email (from a template we provide) to invite that number of the top ranked applicants to sign up for the open training mailing list? 

Caveat: @maneesha and @sheraaron - I know we don't always have the names of learners for member events before the event itself. With this workflow, we would need to either have the learner names 30 days in advance of the event, or we would need to be able to set a field in the event showing how many seats are spoken for. The first option is more difficult administratively, but the second is more difficult programmatically. 

@pbanaszkiewicz - Please let me know if you need any more information or details about how we would like this to be implemented. Also please let us know what the timeline would be for these changes. Thank you.

**Comment by pbanaszkiewicz** [comment]
@ErinBecker:

1. The current [milestone v2.1](https://github.com/swcarpentry/amy/milestone/48) is definitely delayed since I was on vacation this month; I proposed to move v2.1 release to the end of September.
2. The scoring is quite easy to implement, should take a day or two to implement; besides the fields you proposed (`score-auto`, `score-manual`, `score-total`) I'd suggest adding a field for explanation of the value of manual score, just in case.

The bigger issue here is the automation (doing an action at specific time), because it involves other software than usual (and running separately to the web server). This will take probably about 5 days of work for me.

And by days here I don't mean 8-hour work days, because usually I only work in the evenings, so it's about 3 hours per day.

With this and the caveats you mentioned in mind, I'd suggest leaving out this feature and instead allowing human to manage the invites (of course sending out invites would be automatic).

**Comment by ErinBecker** [comment]
@pbanaszkiewicz Thanks for the fast response! I'm glad to hear that this change could make it into the v2.1 release at the end of September. I agree with adding a fourth field (`score-comments`?) for storing comments about the manual portion of the score.

I understand the limitations you mention about having AMY send out emails to people at specific times. Since that might not be a workable solution at this point, could an alternative be:

1) AMY stores applicant scores with each training request
2) It is possible to export a csv of training requests **sorted by `score-total`**
3) 29 days before each training event, @karenword downloads the list of training requests and runs a gmailr script (I can write this) to email a set number of people from the top of the list
4) We would want to somehow note in AMY which applicants have been invited to join - currently, we manually change the `state` tag to `accepted` for each applicant who has been invited. Is there an easier way for us to do this?

I'm not sure what you mean by "sending out invites would be automatic", could you please explain?

Looking forward to getting this workflow confirmed and implemented! It will really help @karenword and the rest of the Instructor Training team (@maneesha and myself) with tracking applicants.

**Comment by pbanaszkiewicz** [comment]
Hello @ErinBecker!

> I'm not sure what you mean by "sending out invites would be automatic", could you please explain?

I meant that steps 3-4 of your comment can be done through AMY - for example via a new special page just for the Instructor Training Team. You can imagine that each TTT event has a "Invite trainees" page that displays training requests sorted by `score-total` in descending order; each request has a checkbox next to it, and top 25-`seats_available` are automatically checked. Before and after the list there would be a button to send invites to the selected trainees OR it would open your email client and prefill BCC & subject & content fields for you.

Additionally, on the main dashboard we can display TTT events that start in the next 30 days and have available seats.

**Comment by ErinBecker** [comment]
@pbanaszkiewicz - That sounds fantastic! Do you need any other info from myself or @karenword @maneesha @sheraaron to get started on implementing this?

**Comment by ErinBecker** [comment]
@pbanaszkiewicz - The Instructor Training team just had a discussion about the logistics of inviting open training applicants to Instructor Training events. Based on our discussion, I wanted to update this issue.

1) The `score_manual`, `score_auto`, `score_total` and `score_notes` fields are the same as described above. 
2) AMY generates a page of applicants sorted by score (as described above)
3) From that page, we can select applicants (with checkboxes) and have the option to: "mark selected as accepted" and "mark selected as rejected". This will change the applicants' training request state to "Accepted" or "Rejected" as appropriate. (Note: Although we don't currently have a "Rejected" state, we should add one if possible as we have been discussing the possibility of rejecting applicants.)
4) From that page we also have the option to "mail all selected". @karenword will use this feature to send an email to newly accepted applicants to let them know they have been accepted and will be receiving future announcements about upcoming Instructor Training events with open seats.
5) **NEW** In the Training Requests page, we already have the ability to filter by State = "Accepted" and Is Matched? = "Unmatched". That generates a list that spans multiple pages and offers the "Mail selected" option at the bottom of the page. However, clicking the box next to "Submitter" at the top of the page only selects the applications on the current page. We want to be able to select all applicants that match the filtering criteria and send an email to them using the "Mail selected" button. Is this possible?

Please let me know if you need any more details. Thanks @pbanaszkiewicz!

**Comment by pbanaszkiewicz** [comment]
@ErinBecker: a new question:
> Award +1 for being in economics or social sciences, arts, humanities, or library science.

Is it +1 when applicant is in economics, or social sciences, or arts/humanities, or library science? Or is it +1 for being in each domain, so up to +4?

**Comment by ErinBecker** [comment]
@pbanaszkiewicz 

Does the `Discarded` status remove the data from AMY? If not, then it will work for our purposes and we can leave these statuses as currently implemented.

For the second question, it is a maximum of +1 for being in one of those fields. They are not additive.

**Comment by karenword** [comment]
I've been using `discarded` status for duplicate applications. While it doesn't seem to lose the information, it also doesn't clearly identify those applications as having been rejected. This may not be crucial information to retain, but worth considering.

**Comment by pbanaszkiewicz** [comment]
@ErinBecker I can confirm what @karenword is saying about information loss.

As for `score_manual`, here's a list of what can and what cannot be computed automatically:

Criteria | Can be automated
----------|-------------------------
Award +1 for being outside Europe/UK/US/Canada/Australia/New Zealand | Only after migration from `django-countries` to `django-countries-plus`, which extends our countries information with continent data (and more)
Award +1 for being in smaller/remote/less affluent institution within EU/UK/US/Can/Aus/NZ | Same as above.
Award +1 for being in economics or social sciences, arts, humanities, or library science. | Yes.
Optionally award +1 for self-identifing as a member of a group that is under-represented in research and/or computing, e.g., women, ethnic minorities, LGBTQ, etc. | No, because the field is textual and needs human to understand it.
Optionally award +1 for being an active contributor to other volunteer or non-profit groups with significant teaching or training components | No, because the field is textual and needs human to understand it.
Award +1 for each previous involvement with The Carpentries up to a maximum bonus of +3. | Yes.
Award +1 for previous training in teaching: "a certification or short course" or "a full degree". | Yes, except for "other" and "explanation".
Award +1 for previous experience in teaching: "TA for full course" or "primary instructor for full course". | Yes, except for "other" and "explanation".
Award +1 for using tools "every day" or "a few times a week". | Yes.
Award +4 for meeting criteria for current recruitment push. | No, unless criteria doesn't change too often.
Award -3 to +3 based on free text responses. | No.

Please go through and decide if this is satisfactory for you, @ErinBecker and @karenword.

**Comment by ErinBecker** [comment]
Criteria | Can be automated
----------|-------------------------
Award +1 for being outside Europe/UK/US/Canada/Australia/New Zealand | Only after migration from `django-countries` to `django-countries-plus`, which extends our countries information with continent data (and more)

Will this be in the upcoming release? Or is it something that is a longer-term migration? If it isn't included in the upcoming release, would we be able to score this automatically if we provided you with a list of countries that should be excluded from this point? (i.e. instead of "Europe" to give you a list of countries.)

Criteria | Can be automated
----------|-------------------------
Award +1 for being in smaller/remote/less affluent institution within EU/UK/US/Can/Aus/NZ | Same as above.

This is a checkbox, so we should be able to score it automatically. Please let me know if I'm misunderstanding. 

Criteria | Can be automated
----------|-------------------------
Award +1 for being in economics or social sciences, arts, humanities, or library science. | Yes.
Optionally award +1 for self-identifing as a member of a group that is under-represented in research and/or computing, e.g., women, ethnic minorities, LGBTQ, etc. | No, because the field is textual and needs human to understand it.

Ok. 

Criteria | Can be automated
----------|-------------------------
Optionally award +1 for being an active contributor to other volunteer or non-profit groups with significant teaching or training components | No, because the field is textual and needs human to understand it.

Ok.

Criteria | Can be automated
----------|-------------------------
Award +1 for each previous involvement with The Carpentries up to a maximum bonus of +3. | Yes.
Award +1 for previous training in teaching: "a certification or short course" or "a full degree". | Yes, except for "other" and "explanation".
Award +1 for previous experience in teaching: "TA for full course" or "primary instructor for full course". | Yes, except for "other" and "explanation".

The Trainers doing the manual scoring can add points if the applicants put something relevant in "other" or "explanation". Only the options you listed above should be given an automatic point.

Criteria | Can be automated
----------|-------------------------
Award +1 for using tools "every day" or "a few times a week". | Yes.
Award +4 for meeting criteria for current recruitment push. | No, unless criteria doesn't change too often.

The criteria will likely change about once every six months (or less frequently). 

Criteria | Can be automated
----------|-------------------------
Award -3 to +3 based on free text responses. | No.

Great. This is what we planned to have Trainers score manually.

**Comment by pbanaszkiewicz** [comment]
@ErinBecker I have one blocker before I start working on the migration from `django-countries` to `django-countries-plus` (I also need to evaluate if we want to migrate, because the latter seems to be less popular and less maintained). If you want to change to countries based evaluation (instead of "EU/UK/US/Can/Aus/NZ"), it's definitely easier to do (no migration; simpler logic); additionally I don't know if `django-countries-plus` maintains list of European Union countries.

> This is a checkbox, so we should be able to score it automatically. Please let me know if I'm misunderstanding.

Well, the checkbox says nothing about EU/UK/US/Can/Aus/NZ. So should I assume that whenever it's "underresourced institution", it should be +1 (OR-ed with criteria 1)?

**Comment by ErinBecker** [comment]
- yes, let's move to country based evaluation using the list in my previous comment

- anytime the box is checked we will do +1. It doesn't need to be "OR-ed". The way the rubric is stated, it seems like it should be "OR-ed", but that's not how we've been scoring. I'll add to our (long) to-do list to update the rubric language. 😄

**Comment by pbanaszkiewicz** [comment]
@ErinBecker do you think [this](https://github.com/swcarpentry/amy/issues/1327#issuecomment-422539917) list of countries is complete? I didn't find Switzerland, Norway or Iceland - while not in EU, these are still rich european countries.

**Comment by ErinBecker** [comment]
@pbanaszkiewicz - you're right. I was using EU membership as a proxy, but we are actually trying to filter out rich countries for this point in the scoring rubric. I will post a new list shortly.

**Comment by ErinBecker** [comment]
Please add to above list:
- Switzerland
- Norway
- Iceland

Remove from the list:
- Bulgaria
- Croatia

The data I am using for this list now comes from World Bank's list of countries by GDP per capita (https://en.wikipedia.org/wiki/List_of_countries_by_GDP_(PPP)_per_capita) with the cutoff set after Latvia (number 48 in the list). 

I am not including countries in Latin America, Brazil, Africa, or Asia in this this list of countries to exclude from the additional point scoring - regardless of whether those countries fall above the GDP per capita cutoff.

**Comment by pbanaszkiewicz** [comment]
Thanks @ErinBecker. Updating the code right now.

**Comment by ErinBecker** [comment]
As discussed at our meeting today (@karenword, @maneesha, @sheraaron, elizabethwilliams8), the changes that we would like to see to the demo @pbanaszkiewicz provided are:

1) Add an "Accept all selected" button to the Training Request page
2) Add filter for people who have not had manual score entered, possibly by setting manual score default to null. This will enable us to download those records for scoring by people who do not have AMY access.

The one other item we discussed will be in a future release, so I will put it in a new issue. Thank you @pbanaszkiewicz!

**Comment by ErinBecker** [comment]
A follow-up question for @pbanaszkiewicz -

When filtering training requests by Matching, there are the following options: 

- unmatched
- matched trainee, unmatched training

In our new workflow that we've been discussing, we want to be able to filter for all training requests that are "Accepted" but "Unmatched". I'm not sure which of the two options above we should be using (or if either of those is actually what we need). Could you please clarify the difference between the two?

**Comment by pbanaszkiewicz** [comment]
Hello @ErinBecker:

> In our new workflow that we've been discussing, we want to be able to filter for all training requests that are "Accepted" but "Unmatched".

In this case, in the filter you should select "State" = "Accepted" and "Is Matched?" = "Unmatched".

In general, requests can be either "Pending", "Discarded" or "Accepted". This you can filter in the "State" field. Training requests, in particular, can be matched with a trainee user account and a training event, and this you can filter with "Is Matched?" field.

Hope this helps!

**Comment by pbanaszkiewicz** [comment]
@ErinBecker I just realized I should clarify one more thing: what is the difference between "Is Matched?": "Unmatched" and "Matched trainee, unmatched training".

Each valid and accepted request has to be connected to existing person account (first), and TTT event (later). The filter "Is Matched?" allows to select all requests that:
1. have no connection to person account nor TTT event (**Unmatched**)
2. have connection to person account, but not to TTT event (**Matched trainee, unmatched training**)
3. have connection to both person account and TTT event (**Matched trainee and training**)

**Comment by ErinBecker** [comment]
🎉 😀
